Controller/Adminhtml/Menu/Edit.php

```diff
@@ -1,58 +1,100 @@
+use Magento\Backend\App\Action\Context;
 use Magento\Framework\Registry;
+use Magento\Framework\View\Result\PageFactory;
 use Snowdog\Menu\Api\MenuRepositoryInterface;
+use Snowdog\Menu\Model\MenuFactory;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\Framework\Controller\ResultInterface;
+use Magento\Backend\Model\View\Result\Redirect;
+use Magento\Backend\Model\View\Result\Page;
 
-class Edit extends Action
+class Edit extends AbstractMenu implements HttpGetActionInterface
 {
     const REGISTRY_CODE = 'snowmenu_menu';
+
     /**
      * @var MenuRepositoryInterface
      */
     private $menuRepository;
+
+    /**
+     * @var PageFactory
+     */
+    private $resultPageFactory;
+
+    /**
+     * @var MenuFactory
+     */
+    private $menuFactory;
+
     /**
      * @var Registry
      */
-    private $registry;
+    private $coreRegistry;
 
-    public function __construct(Action\Context $context, MenuRepositoryInterface $menuRepository, Registry $registry)
-    {
-        parent::__construct($context);
+    /**
+     * @param Context $context
+     * @param Registry $coreRegistry
+     * @param PageFactory $resultPageFactory
+     * @param MenuRepositoryInterface $menuRepository
+     * @param MenuFactory $menuFactory
+     */
+    public function __construct(
+        Context $context,
+        Registry $coreRegistry,
+        PageFactory $resultPageFactory,
+        MenuRepositoryInterface $menuRepository,
+        MenuFactory $menuFactory
+    ) {
+        $this->resultPageFactory = $resultPageFactory;
         $this->menuRepository = $menuRepository;
-        $this->registry = $registry;
+        $this->menuFactory = $menuFactory;
+        $this->coreRegistry = $coreRegistry;
+        parent::__construct(
+            $context
+        );
     }
 
     /**
-     * Dispatch request
-     *
-     * @return \Magento\Framework\Controller\ResultInterface|ResponseInterface
-     * @throws \Magento\Framework\Exception\NotFoundException
+     * @SuppressWarnings(PHPMD.NPathComplexity)
      */
-    public function execute()
+    public function execute(): ResultInterface
     {
-        $id = $this->getRequest()->getParam('id');
-        try {
-            $model = $this->menuRepository->getById($id);
-            $this->registry->register(self::REGISTRY_CODE, $model);
-            $result = $this->resultFactory->create(ResultFactory::TYPE_PAGE);
-            $result->setActiveMenu('Snowdog_Menu::menus');
-            $result->getConfig()->getTitle()->prepend(__('Edit Menu %1', $model->getTitle()));
-            return $result;
-        } catch (NoSuchEntityException $e) {
-            $result = $this->resultRedirectFactory->create();
-            $result->setPath('*/*/index');
-            return $result;
+        $menuId = (int) $this->getRequest()->getParam(self::ID);
+
+        if ($menuId) {
+            try {
+                $model = $this->menuRepository->getById($menuId);
```

Comment:
Maybe we should move this part to `\Snowdog\Menu\Controller\Adminhtml\AbstractMenu` as `initMenu` method? I see it is used in other places too.

Reply:
How did you suppose to move it to initMenu method?

Reply:
I think Darek meant to add it to an abstract class, so it can be called from there in children classes, something like this example within Magento 2 core: 

Method definition: https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Customer/Controller/Adminhtml/Index.php#L259
Method call: https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Customer/Controller/Adminhtml/Index/Cart.php#L157

Reply:
I used `getCurrentMenu()` method here as init method would do basically the same.